### PR TITLE
#396 Add ReportData class

### PR DESCRIPTION
**Pull request by HDouss**
Resolving issue #396

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#401](https://github.com/yegor256/jpeek/issues/401) is assigned to @fanifieiev[/z](https://www.0crat.com/u/fanifieiev), here is [why](https://www.0crat.com/footprint/C7JGJ00DP/559c6f4c-981f-4796-984d-254594550023); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

**Comment by HDouss** [comment]
@fanifieiev ping

**Comment by HDouss** [comment]
@fanifieiev ping...

**Comment by HDouss** [comment]
@fanifieiev ping...again..

**Comment by HDouss** [comment]
@fanifieiev ping...

File: src/test/java/org/jpeek/ReportDataTest.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Test
+    public void reportsData() throws Exception {
+        final String name = "whatever";
+        final Random random = new Random();
+        final double mean = random.nextDouble();
+        final double sigma = random.nextDouble();
+        final Map<String, Object> sample = ReportDataTest.args();
+        final ReportData data = new ReportData(name, sample, mean, sigma);
+        new Assertion<>("Must returns name", data.metric(), new IsEqual<>(name)).affirm();
```
Comment (fanifieiev) [inline src/test/java/org/jpeek/ReportDataTest.java]: @HDouss It is better to reformat here and below in a way:
```
new Assertion<>(
   "<message>",
   ... ,
   ...
).affirm();
```
for better readability

File: src/main/java/org/jpeek/ReportData.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    public double sigma() {
+        return this.sig;
+    }
+
+    /**
+     * Params accessor.
+     * @return Params
+     */
+    public Map<String, Object> params() {
+        return new HashMap<>(this.args);
```
[the remark this answers is not in the document]
Comment (HDouss) [reply src/main/java/org/jpeek/ReportData.java]: @fanifieiev Ok. but I think I should remove the test for immutability or change it to expect Exceptions.

[the remark this answers is not in the document]
Comment (fanifieiev) [reply src/main/java/org/jpeek/ReportData.java]: @HDouss I would personally not create a test for `ReportData`, it is just a holder object, just dummy object, I would better check how much it is involved in XslReportTest

[the remark this answers is not in the document]
Comment (HDouss) [reply src/main/java/org/jpeek/ReportData.java]: @fanifieiev it does not harm to have such tests, especially for immutability.

**Comment by HDouss** [comment]
@paulodamaso ping

File: src/main/java/org/jpeek/ReportData.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+     * @param name Name of the metric
+     * @param args Params for XSL
+     * @param mean Mean
+     * @param sigma Sigma
+     * @checkstyle ParameterNumberCheck (10 lines)
+     */
+    ReportData(final String name, final Map<String, Object> args, final double mean,
+        final double sigma) {
+        this.metr = name;
+        this.args = Collections.unmodifiableMap(new HashMap<>(args));
```
Comment (paulodamaso) [inline src/main/java/org/jpeek/ReportData.java]: @HDouss Let's avoid using static methods. See if we can't replace this call by some of the `cactoos` collections

File: src/test/java/org/jpeek/ReportDataTest.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+public final class ReportDataTest {
+
+    @Test
+    public void reportsData() throws Exception {
+        final String name = "whatever";
+        final double mean = 0;
+        final double sigma = 1;
+        final Map<String, Object> sample = ReportDataTest.args();
+        final ReportData data = new ReportData(name, sample, mean, sigma);
+        new Assertion<>(
```
Comment (paulodamaso) [inline src/test/java/org/jpeek/ReportDataTest.java]: @HDouss Please split this method into several methos containing just one assertion each

File: src/test/java/org/jpeek/ReportDataTest.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @Test
+    public void shouldBeImmutable() throws Exception {
+        final String name = "metric";
+        final Map<String, Object> sample = ReportDataTest.args();
+        final Map<String, Object> params = new HashMap<>(sample);
+        final ReportData data = new ReportData(name, params);
+        params.clear();
+        new Assertion<>(
```
Comment (paulodamaso) [inline src/test/java/org/jpeek/ReportDataTest.java]: @HDouss Here also, please split it into two methods

**Comment by HDouss** [comment]
@paulodamaso Thanks. Fixed.

**Comment by HDouss** [comment]
@paulodamaso ping

**Comment by 0crat** [comment]
@fanifieiev[/z](https://www.0crat.com/u/fanifieiev) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by HDouss** [comment]
@paulodamaso ping.

**Comment by 0crat** [comment]
Job audit: `ARC` didn't merge or close PR, performer @fanifieiev[/z](https://www.0crat.com/u/fanifieiev)[/z](https://www.0crat.com/u/fanifieiev) didn't make CR comments

**Pull request merged**